python/cuspatial/cuspatial/core/binpreds/binpreds.py

```diff
@@ -234,3 +267,233 @@ def preprocess(self, lhs, rhs):
+        # Save temporary values since x cannot be used unmodified
+        x = linestring.x
+        y = linestring.y
+        point_range = cudf.Series(cp.arange(len(x)))
+        indices = point_range.groupby(linestring.point_indices())
+        # Create masks for the first and last values of each linestring
+        swap_x = x[indices.last()].reset_index(drop=True) < x[
+            indices.first()
+        ].reset_index(drop=True)
+        swap_y = y[indices.last()].reset_index(drop=True) < y[
+            indices.first()
+        ].reset_index(drop=True)
+        # Swap the first and last values of each linestring
+        (x.iloc[indices.last()[swap_x]], x.iloc[indices.first()[swap_x]],) = (
+            x.iloc[indices.first()[swap_x]],
+            x.iloc[indices.last()[swap_x]],
+        )
+        (y.iloc[indices.last()[swap_y]], y.iloc[indices.first()[swap_y]],) = (
+            y.iloc[indices.first()[swap_y]],
+            y.iloc[indices.last()[swap_y]],
+        )
+        return cudf.Series(cp.column_stack((x.values, y.values)))
+
+    def _sort_linestrings(self, lhs, rhs, initial):
+        """Swap first and last values of each linestring to ensure that
+        the first point is the lowest value. This is necessary to ensure
+        that the endpoints are not included in the comparison."""
+        # TODO: Refactor the 4x repetition here into a utility method.
+
+        lhs_xy = self._swap_linestring_first_and_last(lhs)
+        rhs_xy = self._swap_linestring_first_and_last(rhs)
+
+        return (
+            PreprocessorOutput(lhs_xy, lhs.point_indices()),
+            PreprocessorOutput(rhs_xy, rhs.point_indices()),
+            initial,
+        )
+
+    def _polygons_to_multipoints(self, polygons):
+        """Strip the first value of each polygon"""
+        offsets = polygons.ring_offset[:-1]
+        x = polygons.x
+        x[offsets] = None
+        x = x.dropna()
+        y = polygons.y
+        y[offsets] = None
+        y = y.dropna()
+        new_offsets = self.lhs.polygons.ring_offset - cp.arange(
+            len(self.lhs.polygons.ring_offset)
+        )
+        xy = cudf.DataFrame({"x": x, "y": y}).interleave_columns()
+        return cuspatial.GeoSeries.from_multipoints_xy(
+            xy, new_offsets.astype("int32")
+        )
+
+    def _sort_polygons(self, lhs, rhs, initial):
+        """Sort xy according to bins defined by offset"""
+
+        lhs_polygons_as_multipoints = self._polygons_to_multipoints(lhs)
+        rhs_polygons_as_multipoints = self._polygons_to_multipoints(rhs)
+
+        return self._sort_multipoints(
+            lhs_polygons_as_multipoints.multipoints,
+            rhs_polygons_as_multipoints.multipoints,
+            initial,
+        )
+
+    def preprocess(self, lhs, rhs):
+        # Compare types
+        type_compare = lhs.dtype == rhs.dtype
+        # Any unmatched type is not equal
+        if (type_compare == False).all():  # noqa: E712
```

Comment:
Because `lhs.dtype` returns a `Series` of dtypes, one for each row in the `GeoSeries`.

Reply:
I've never seen a `dtype` property that returns a series instead of a scalar... This is new to me. Would you mind point me to where this property is documented?

Reply:
`GeoSeries.py:103`, it just returns the `self._column._meta_input_types`.